### PR TITLE
Fix and simplify status icon indicator

### PR DESCRIPTION
**Pull request by flozia**
## Description

Fixes and simplifies the status indicator color indicator.

## Reference

- #4364
- #4366 

## Checklist
    
- [x] My code follows the style guidelines for this project
- [x] I have not added any packages that contain high risk or unknown licenses (GPL,  LGPL, MPL, etc. consult with DevOps if in question)
- [x] I have performed a self review of my own code
- [x] I have commented my code PARTICULARLY in hard to understand areas
- [x] I have added thorough tests where needed

### CODE REVIEW

File: src/statusicon.cpp [hunk 1]
```diff
@@ -67,38 +72,41 @@ void StatusIcon::activateAnimation() {
...
-const QIcon& StatusIcon::icon() {
-  if (m_icon.isNull()) {
-    stateChanged();
-    Q_ASSERT(!m_icon.isNull());
+      m_indicatorColor = QColor();
+      break;
   }
 
-  return m_icon;
+  return m_indicatorColor;
```
Comment (bakulf) [inline src/statusicon.cpp:109]: Why do we need this variable?

File: src/statusicon.cpp [hunk 2]
```diff
@@ -176,29 +172,14 @@ QIcon StatusIcon::drawStatusIndicator(const QString& iconUrl) const {
...
+void StatusIcon::setIcon(const QString& iconUrl) {
   logger.debug() << "Set icon" << iconUrl;
 
-  m_iconUrl = iconUrl;
-
-  // If we are in state main and the VPN is on we draw the status indicator
-  // directly onto the image Windows and Linux.
-  if (shouldDrawIndicator) {
+  // Draw the status indicator directly onto the image if needed
+  if (!m_icon.isNull()) {
```
Comment (bakulf) [inline src/statusicon.cpp:179]: Here you can simply invalidate m_icon

File: src/statusicon.cpp [hunk 3]
```diff
@@ -48,17 +50,20 @@ StatusIcon::StatusIcon() : m_iconUrl(LOGO_GENERIC) {
...
-          [this] { setIcon(m_iconUrl, false); });
-#endif
 }
 
 StatusIcon::~StatusIcon() { MVPN_COUNT_DTOR(StatusIcon); }
 
+const QIcon& StatusIcon::icon() {
+  if (m_icon.isNull()) {
+    // Inits m_icon according to the current state
+    stateChanged();
```
Comment (bakulf) [inline src/statusicon.cpp:60]: m_icon = draw()

File: src/statusicon.cpp [hunk 4]
```diff
@@ -67,38 +72,41 @@ void StatusIcon::activateAnimation() {
...
-    Q_ASSERT(!m_icon.isNull());
+      m_indicatorColor = QColor();
+      break;
   }
 
-  return m_icon;
+  return m_indicatorColor;
 }
 
 void StatusIcon::stateChanged() {
```
Comment (bakulf) [inline src/statusicon.cpp:112]: Here you can simply invalidate the m_icon and emit the signal. 
If you do this, you can remove `setIcon`

File: src/statusicon.h [hunk 5]
```diff
@@ -16,37 +16,32 @@ class StatusIcon final : public QObject {
...
-
  public:
   StatusIcon();
   ~StatusIcon();
 
-  QUrl iconUrl() const;
-  const QString& iconString() const { return m_iconUrl; }
   const QIcon& icon();
-  const QColor& indicatorColor() const { return m_indicatorColor; }
+  const QString& iconString() const { return m_iconUrl; }
```
Comment (bakulf) [inline src/statusicon.h:24]: Compute the iconString at each call. And remove m_iconUrl

File: src/statusicon.h [hunk 6]
```diff
@@ -16,37 +16,32 @@ class StatusIcon final : public QObject {
...
   QIcon drawStatusIndicator(const QString& iconUrl) const;
-  void setIcon(const QString& iconUrl, bool shouldDrawIndicator);
-  void setIndicatorColor(const QColor& color);
+  void generateIcon();
+  void setIcon(const QString& iconUrl);
 
  private:
-  QString m_iconUrl;
   QIcon m_icon;
+  QString m_iconUrl;
```
Comment (bakulf) [inline src/statusicon.h:44]: You can remove this or to optimize the code, keep it and use it as a cache like m_icon.
In stateChanged, set it to null.

File: src/statusicon.h [hunk 7]
```diff
@@ -16,37 +16,32 @@ class StatusIcon final : public QObject {
...
   QColor m_indicatorColor;
```
Comment (bakulf) [inline src/statusicon.h:45]: And this

File: tests/unit/teststatusicon.cpp [hunk 8]
```diff
@@ -10,64 +10,49 @@
...
 #else
-  QCOMPARE(si.iconUrl().toString(), "qrc:/ui/resources/logo-generic-mask.png");
   QCOMPARE(si.iconString(), ":/ui/resources/logo-generic-mask.png");
 #endif
 
   TestHelper::vpnState = MozillaVPN::StateMain;
   TestHelper::controllerState = Controller::StateOn;
   si.stateChanged();
 #if defined(MVPN_LINUX) || defined(MVPN_WINDOWS)
-  QCOMPARE(si.iconUrl().toString(), "qrc:/ui/resources/logo-on.png");
```
Comment (bakulf) [inline tests/unit/teststatusicon.cpp:33]: Why this change? Here we can probably remove all of these #ifdef macro

**Review by bakulf — CHANGES_REQUESTED** [review]

File: src/statusicon.cpp (hunk 1 above)
Comment (flozia) [reply src/statusicon.cpp:109]: We do not need it and can return the colors directly.

File: src/statusicon.cpp [hunk 9]
```diff
@@ -43,117 +46,122 @@ constexpr const char* LOGO_GENERIC_ON =
...
     default:
-      setIcon(LOGO_GENERIC, false);
+      m_animatedIconTimer.stop();
+      return LOGO_GENERIC;
       break;
   }
 }
 
-QIcon StatusIcon::drawStatusIndicator(const QString& iconUrl) const {
+const QColor& StatusIcon::indicatorColor() {
```
Comment (bakulf) [inline src/statusicon.cpp:125]: I would not return a reference. Return a QColor

File: src/statusicon.cpp [hunk 10]
```diff
@@ -43,117 +46,122 @@ constexpr const char* LOGO_GENERIC_ON =
...
+      return RED_COLOR;
+    default:
+      logger.error() << "Unhandled status indicator for connection stability";
+      return INVALID_COLOR;
+  }
+
+  return m_indicatorColor;
+}
+
+void StatusIcon::stateChanged() {
```
Comment (bakulf) [inline src/statusicon.cpp:151]: What about if we call this method `refreshNeeded` or `resetIcon`? Asking because it's not connected to state changing anymore.

File: tests/unit/teststatusicon.cpp [hunk 11]
```diff
@@ -10,64 +10,49 @@
 void TestStatusIcon::basic() {
   StatusIcon si;
 #if defined(MVPN_LINUX) || defined(MVPN_WINDOWS)
-  QCOMPARE(si.iconUrl().toString(), "qrc:/ui/resources/logo-generic.png");
```
Comment (bakulf) [inline tests/unit/teststatusicon.cpp:13]: You can still keep these tests

**Review by bakulf — CHANGES_REQUESTED** [review]
Great refactoring! Can you please improve the tests? I think you can remove the ifdef and test the single methods on all the platforms.

File: src/statusicon.cpp (hunk 10 above)
Comment (flozia) [reply src/statusicon.cpp]: Sounds good!

File: tests/unit/teststatusicon.cpp (hunk 11 above)
Comment (flozia) [reply tests/unit/teststatusicon.cpp:13]: The method `iconUrl` is not used anymore.

**Review by bakulf — APPROVED** [review]
